### ISSUE TITLE
PUT operations on existing Facebook authenticated users resulting in 404s

### ISSUE THREAD

**Issue by ShinyMark** [ParsePlatform/parse-server#1422]
#### Environment Setup

- parse-server 2.2.6 running on Heroku
- Database hosted on mLab, was migrated from Parse.com
- .NET Parse client 1.7 running under Unity 5.3.4
- Client is running on Android/iOS 

#### Problem

When the client app tries to update a _User in the database via a PUT parse-server rejects it with a 404. I've searched for other issues posted previously and saw mention of an "authData" header but I'm not seeing it in my requests.

Doing a GET on the same _User object works fine with a 200 and returns the expected data.

Is this an ACL problem? Or is the client not sending some auth header it is supposed to? The expected access rights in our app is:

1. Any user can read any other user's _User object.
2. Only you can write/update your own _User object.

If I remember correctly that's the default. All of our users were created via Facebook login. That's the only identity authentication method we support. Our app is behaving mostly as expected after moving to parse-server, so I'm certain read permissions for _User objects are working, but the logged in user cannot update their own _User object.

#### Logs/Trace

With VERBOSE=1...

*Request*

`verbose: PUT /parse/classes/_User/<redacted> { host: '<redacted>.herokuapp.com',
  connection: 'close',
  'content-type': 'application/json',
  'x-unity-version': '5.3.4f1',
  accept: '*/*',
  'x-parse-session-token': '<redacted>',
  'x-parse-application-id': '<redacted>',
  'x-parse-windows-key': '<redacted>',
  'x-parse-installation-id': '<redacted>',
  'x-parse-os-version': 'iPhone7,1',
  'accept-language': 'en-us',
  'accept-encoding': 'gzip, deflate',
  'x-parse-client-version': 'net-unity1.7.0.0',
  'user-agent': '<redacted>/0 CFNetwork/758.3.15 Darwin/15.4.0',
  'x-parse-app-build-version': '0',
  'x-parse-app-display-version': '<redacted>',
  'x-request-id': '4fc2d0e7-d946-411e-8a08-fa70d91da319',
  'x-forwarded-for': '<redacted>',
  'x-forwarded-proto': 'https',
  'x-forwarded-port': '443',
  via: '1.1 vegur',
  'connect-time': '0',
  'x-request-start': '1460075515384',
  'total-route-time': '0',
  'content-length': '2270' }`

I removed the app-specific body of the object.

*Result*

`heroku[router]: at=info method=POST path="/parse/classes/_User/<redacted>" host=<redacted>.herokuapp.com request_id=4fc2d0e7-d946-411e-8a08-fa70d91da319 fwd="<redacted>" dyno=web.1 connect=0ms service=389ms status=404 bytes=546`
`app[web.1]: verbose: error: code=101, message=Permission denied for this action.`

**Comment by flovilmart** [comment]
Permission denied for this action would be a locked down Class Level Permission. can you check if you have any on _User ?

**Comment by ShinyMark** [comment]
Thank you for the quick response. I'm a bit of a newbie here in terms of how Parse works behinds the scenes so I may need some more guidance on where to look.

I'm looking around in mLab at our data now. Should I be looking at _SCHEMA? I see a "class_permissions" section for _User in there as well as a per-field schema defining the data types in our _User documents. Here's the contents of class_permissions.

`        "class_permissions": {
            "get": {
                "*": true
            },
            "find": {
                "*": true
            },
            "update": {
                "*": true
            },
            "create": {
                "*": true
            },
            "delete": {
                "*": true
            },
            "addField": {},
            "readUserFields": [],
            "writeUserFields": []
        }
`

I may be misunderstanding but this looks somewhat scary. Does "*" for delete means anyone can delete anyone else's _User document? I hope not!

I found the specific _User document being tested and I don't see any access control fields within the document. I do see `_auth_data_facebook` etc. 

Is there any other information I can provide?

**Comment by flovilmart** [comment]
that sounds good, (I edited my comment) and it seems to be an ACL problem, what does it look like on that user of yours that's failing to update?

**Comment by flovilmart** [comment]
maybe related to  #1192

**Comment by ShinyMark** [comment]
I stripped every piece of data out of the _User record that is specific to our app. This is what was left.

`{
    "_id": "<redacted>",
    "_auth_data_facebook": {
        "access_token": "<redacted>",
        "expiration_date": {
            "$date": "3000-01-01T00:00:00.000Z"
        },
        "id": "<redacted>"
    },
    "_created_at": {
        "$date": "2014-06-08T04:02:16.397Z"
    },
    "_hashed_password": "<redacted>",
    "_perishable_token": "<redacted>",
    "_session_token": "<redacted>",
    "_updated_at": {
        "$date": "2016-04-07T16:41:33.563Z"
    }, 
    "username": "<redacted>"
}`

Nothing related to access control as far as I can see.

**Comment by flovilmart** [comment]
Do you set ACL like publicReadAccess false, publicWriteAccess false?

**Comment by ShinyMark** [comment]
Nope. The client code uses ParseFacebookUtils.LogInAsync(). After that there are zero ACL modifications requested by the client and we have no cloud code.

**Comment by ShinyMark** [comment]
Ok, we figured this out. The problem was that we had set the _SCHEMA table to disable addField per best practices for a live product. A developer on the team added a new piece of data to _User in the client but didn't explicitly add it in the backend. So when the client tried to update the _User document with a new field it was correctly rejected.

My only request to the parse-server contributors would be to make the error messaging here more obvious. Perhaps adding an additional sentence to the existing message like "failed to add field <xyz> due to lack of permission in _SCHEMA."

Thanks for the help @flovilmart.
